nsot/api/views.py

```diff
@@ -613,6 +613,85 @@ def networks(self, request, pk=None, site_pk=None, *args, **kwargs):
 
         return self.list(request, queryset=interface.networks, *args, **kwargs)
 
+    @detail_route(methods=['get'])
+    def circuit(self, request, pk=None, site_pk=None, *args, **kwargs):
+        """Return the Circuit I am associated with"""
+        interface = self.get_resource_object(pk, site_pk)
+        cir_a = interface.circuits_a.all()
+        if cir_a:
+            cir_a = serializers.CircuitSerializer(cir_a[0])
+            return self.success(cir_a.data)
+
+        cir_z = interface.circuits_z.all()
+        if cir_z:
+            cir_z = serializers.CircuitSerializer(cir_z[0])
+            return self.success(cir_z.data)
+
+        msg = 'No Circuit found at Interface (site_id, id) = ({}, {})'.format(
+            site_pk,
+            pk,
+        )
+        self.not_found(pk, msg=msg)
+
+
+class CircuitViewSet(ResourceViewSet):
+    """
+    API endpoint that allows Circuits to be viewed or edited.
+    """
+    queryset = models.Circuit.objects.all()
+    serializer_class = serializers.CircuitSerializer
+    filter_class = filters.CircuitFilter
+
+    @cache_response(cache_errors=False, key_func=cache.list_key_func)
+    def list(self, *args, **kwargs):
+        """Override default list so we can cache results."""
+        return super(CircuitViewSet, self).list(*args, **kwargs)
+
+    @cache_response(cache_errors=False, key_func=cache.object_key_func)
+    def retrieve(self, *args, **kwargs):
+        """Override default retrieve so we can cache results."""
+        return super(CircuitViewSet, self).retrieve(*args, **kwargs)
+
+    def get_serializer_class(self):
+        if self.request.method == 'POST':
+            return serializers.CircuitCreateSerializer
+        if self.request.method == 'PUT':
+            return serializers.CircuitUpdateSerializer
+        if self.request.method == 'PATCH':
+            return serializers.CircuitPartialUpdateSerializer
+
+        return self.serializer_class
+
+    @detail_route(methods=['get'])
+    def interfaces(self, request, pk=None, site_pk=None, *args, **kwargs):
+        """Return a list of interfaces for this Circuit"""
+        circuit = self.get_resource_object(pk, site_pk)
+        interfaces = circuit.interfaces
+
+        return self.list(request, queryset=interfaces, *args, **kwargs)
+
+    @detail_route(methods=['get'])
```

Comment:
For `.addresses()` and `.devices()` detail routes the underlying queries should be defined as model methods/properties on the `Circuit` model, and then called from the view methods. We should try to prevent anything other than basic presentation logic being in the views.